Summarize word emb

Lundez opened the pull request.

Lundez committed:
First implementation complete

- Contains Rosellio scoring method
- Contains Ghalandari scoring method
- EmbeddingSummarizer support % & #.
- TODO A little refactoring
- TODO Enums for scoringMethod not working currently (should default to Ghalandari)
- TODO Need to remove obj-test

Lundez committed:
Updated README & minor refactoring

Lundez commented on src/main/kotlin/com/londogard/summarize/embeddings/WordEmbeddings.kt:
```diff
@@ -198,22 +197,7 @@ fun <T> PriorityQueue<T>.addR(element: T): PriorityQueue<T> {
-    }
-}
\ No newline at end of file
+fun sumVector(vectors: List<Array<Float>>): Array<Float> = vectors.reduce { agg, itr -> agg + itr }
```
`private`

denkhan replied:
Would be nice with better names in reduce?

Lundez commented on src/main/kotlin/com/londogard/summarize/summarizers/EmbeddingClusterSummarizer.kt:
```diff
@@ -1,3 +1,166 @@
+) : SmileOperators, SummarizerModel {
+    private val embeddings: WordEmbeddings = WordEmbeddings(dimensions = 50)
+    private val zeroArray = Array(embeddings.dimensions) { 0f }
+    private fun String.simplify(): String =
```
Could probably be a one-liner, let's see how it looks in the IDE

Lundez replied:
And is the `joinToString` really correct..? I think not.

Lundez replied:
Yeah, we want to split `it's` to `it 's` if possible :)

Lundez commented on src/main/kotlin/com/londogard/summarize/summarizers/EmbeddingClusterSummarizer.kt:
```diff
@@ -1,3 +1,166 @@
+    private fun getTfIdf(sentences: List<String>): Pair<List<Double>, Map<Int, String>> {
+        val corpus = sentences.map { it.bag(stemmer = null) }
+        val words = corpus.flatMap { bag -> bag.keys }.distinct()
+        val bags = corpus.map {
```
one-line & put named lambda. 
Finally, it's correct as the words vectorizing the sentence is the full corpora

Lundez commented on src/main/kotlin/com/londogard/summarize/summarizers/EmbeddingClusterSummarizer.kt:
```diff
@@ -1,3 +1,166 @@
+        } // Is this really how to do it? Shouldn't we do one vectorization in total per document rather than sent?
+        val vectors = tfidf(bags)
+        val vector = vectors.mutableSumByCols()
+        return (vector.max()?.let { max -> List(vector.size) { i -> vector[i] / max } }
```
Could we simplify this?

denkhan replied:
Kanske går att använda sig utav map?

Lundez replied:
Vi kan ta bort hela skiten :smile:  Använder bara den för att få wordOfInterest

Lundez commented on src/main/kotlin/com/londogard/summarize/summarizers/EmbeddingClusterSummarizer.kt:
```diff
@@ -1,3 +1,166 @@
+            }
+    }
+
+    override fun summarize(text: String, lines: Int): String {
```
**OBS** The scoringConfig is not used. Update! :warning:

Lundez commented on src/main/kotlin/com/londogard/summarize/summarizers/EmbeddingClusterSummarizer.kt:
```diff
@@ -1,3 +1,166 @@
+    override fun summarize(text: String, ratio: Double): String {
+        val lines = text.sentences().size
+
+        return summarize(text, (lines * ratio).roundToInt().coerceAtLeast(1))
```
This `(lines * ratio)` is used in more than one place, perhaps extract?

denkhan replied:
Nah

Lundez commented on src/main/kotlin/com/londogard/summarize/summarizers/EmbeddingClusterSummarizer.kt:
```diff
@@ -1,3 +1,166 @@
+        return summarize(text, (lines * ratio).roundToInt().coerceAtLeast(1))
+    }
+
+    data class ScoreHolder(val i: Int, val rawSentence: String, val score: Double, val vector: Array<Float>) {
```
Extract into its own file imo

Lundez commented on src/main/kotlin/com/londogard/summarize/summarizers/SummarizerModel.kt:
```diff
@@ -1,6 +1,6 @@
-interface Summarizer {
+interface SummarizerModel {
     fun summarize(text: String, lines: Int = 10): String
     fun summarize(text: String, ratio: Double = 0.1): String
```
Default 0.3 makes more sense.

Lundez reviewed (commented):
Mini-review

denkhan commented on src/main/kotlin/com/londogard/summarize/embeddings/WordEmbeddings.kt:
```diff
@@ -170,26 +166,22 @@ class WordEmbeddings(
                     }
                 }
             }?.reversed()
     }
```
Kanske bättre namn på agg och itr?

Lundez replied:
Makes sense, but agg = aggregator and is good to keep imo.
Itr aka iterator could be renamed into what it is (vector) or such 👍

denkhan commented on src/main/kotlin/com/londogard/summarize/summarizers/EmbeddingClusterSummarizer.kt:
```diff
@@ -13,11 +13,11 @@ enum class ScoringConfig {
+) : SmileOperators, SummarizerModel {
     private val embeddings: WordEmbeddings = WordEmbeddings(dimensions = 50)
     private val zeroArray = Array(embeddings.dimensions) { 0f }
     private fun String.simplify(): String =
```
Något jag inte ser eller slutar raden med "=" ?

Lundez replied:
Something you're not seeing. Making it into a one-liner

denkhan commented on src/main/kotlin/com/londogard/summarize/Extensions.kt:
```diff
@@ -0,0 +1,9 @@
+        for (i in columns.indices)
+            columnSum[i] += columns[i]
+    return columnSum.toList()
+}
```
Finns sumByDouble(), kanske skulle kunna va något? Annars e de ju rätt så lättläst ändå

Lundez replied:
`sumByDouble` räknar ut summan för en vektor. 
`mutableSumByColumns` räknar ut summan för varje vektor. Ex:
`listOf(1,2).sumByDouble() = 3`
`listOf(listOf(1,2), listOf(3,1)).mutableSumByCol() = listOf(4, 3)`

denkhan commented on src/main/kotlin/com/londogard/summarize/embeddings/WordEmbeddings.kt:
```diff
@@ -76,6 +76,11 @@ class WordEmbeddings(
+        if (v1.any { it > 0 } && v2.any { it > 0 }) {
+            (2 - cosine(v1, v2)) / 2.0
+        } else 0.0
+
```
Förstår inte riktigt vad den gör. Men vad händer om v1 = [-0.8,0,8]? alltså ett postivt och ett negativt tal?

Lundez replied:
Tog bort denna, kör direkt på `cosine` istället. Tror denna används för att enbart få positiva värden, men inte hundra

denkhan commented on src/main/kotlin/com/londogard/summarize/extensions/FloatArray.kt:
```diff
@@ -1,16 +1,17 @@
 
 fun Array<Float>.normalize(): Array<Float> {
-    val l = 1f / this.size
+    val l = if (size > 0) (1f / size) else 1f
```
Hur reagerar kotlin på size ? 1f/size : 1f

Lundez replied:
Samma som allt, X / 0 = crash.
Ändrade till äkta normalisering istället (sqrt(sum(xi^2)))

Lundez committed:
Fixed review comments

- Normalize is now a "true" normalisation (i.e. sqrt(sum(xi²)))
- ScoringConfig works
- WordEmbedding now correctly sortByDesc to have top = best match
- Better default values
- ScoreHolder extracted into its own file
- Minor fixes

Lundez committed:
Fixed final review comment that was missed

Lundez committed:
Minor improvements
